lib/chef/dsl/declare_resource.rb

```diff
@@ -297,13 +297,15 @@ def declare_resource(type, name, created_at: nil, run_context: self.run_context,
       #     action :delete
       #   end
       #
-      def build_resource(type, name, created_at: nil, run_context: self.run_context, &resource_attrs_block)
+      def build_resource(type, name, created_at: nil, run_context: self.run_context, enclosing_provider: nil, &resource_attrs_block)
         created_at ||= caller[0]
 
         # this needs to be lazy in order to avoid circular dependencies since ResourceBuilder
         # will requires the entire provider+resolver universe
         require_relative "../resource_builder" unless defined?(Chef::ResourceBuilder)
 
+        enclosing_provider ||= self if is_a?(Chef::Provider)
```

Comment:
maybe a comment on how this is used with the spec helper only

Reply:
as its generic though we can use the same trick elsewhere

Reply:
(and its not just for the spec helper, the enclosing_provider goes back to chef-11-era)

Reply:
If we're going to use it elsewhere then go for it. It just seems like passing it right now is a trick specifically for the specs. The docs on the spec side are solid for that which is probably fine.